sdks/go/pkg/beam/testing/teststream/teststream.go

```diff
@@ -0,0 +1,131 @@
+		return fmt.Errorf("watermark must be monotonally increasing, is at %v, got %v", c.Watermark, timestamp)
+	}
+	watermarkAdvance := &pipepb.TestStreamPayload_Event_AdvanceWatermark{NewWatermark: timestamp}
+	watermarkEvent := &pipepb.TestStreamPayload_Event_WatermarkEvent{WatermarkEvent: watermarkAdvance}
+	c.Events = append(c.Events, &pipepb.TestStreamPayload_Event{Event: watermarkEvent})
+	c.Watermark = timestamp
+	return nil
+}
+
+// AdvanceWatermarkToInfinity advances the watermark for the PCollection corresponding to the tag
+// to the maximum timestamp.
+func (c *Config) AdvanceWatermarkToInfinity() error {
+	return c.AdvanceWatermark(mtime.MaxTimestamp.Milliseconds())
+}
+
+// AdvanceProcessingTime adds an event into the Config Events struct advancing the processing time by a given
+// duration. This advancement is applied to all of the PCollections output by the TestStream.
+func (c *Config) AdvanceProcessingTime(duration int64) {
+	processingAdvance := &pipepb.TestStreamPayload_Event_AdvanceProcessingTime{AdvanceDuration: duration}
+	processingEvent := &pipepb.TestStreamPayload_Event_ProcessingTimeEvent{ProcessingTimeEvent: processingAdvance}
+	c.Events = append(c.Events, &pipepb.TestStreamPayload_Event{Event: processingEvent})
+}
+
+// AdvanceProcessingTimeToInfinity moves the TestStream processing time to the largest possible
+// timestamp.
+func (c *Config) AdvanceProcessingTimeToInfinity() {
+	c.AdvanceProcessingTime(mtime.MaxTimestamp.Milliseconds())
+}
+
+// AddElements adds a number of elements to the Config object at the specified timestamp.
+// The encoder will panic if there is a type mismatch between the provided coder and the
+// elements.
+func (c *Config) AddElements(timestamp int64, elements ...interface{}) error {
+	newElements := []*pipepb.TestStreamPayload_TimestampedElement{}
+	enc := beam.NewElementEncoder(c.ElmCoder.T.Type())
+	for _, e := range elements {
+		var buf bytes.Buffer
+		if err := enc.Encode(e, &buf); err != nil {
+			return fmt.Errorf("encoding value %v failed, got %v", e, err)
+		}
+		newElements = append(newElements, &pipepb.TestStreamPayload_TimestampedElement{EncodedElement: buf.Bytes(), Timestamp: timestamp})
+	}
+	addElementsEvent := &pipepb.TestStreamPayload_Event_AddElements{Elements: newElements}
+	elementEvent := &pipepb.TestStreamPayload_Event_ElementEvent{ElementEvent: addElementsEvent}
+	c.Events = append(c.Events, &pipepb.TestStreamPayload_Event{Event: elementEvent})
+	return nil
+}
+
+// TestStream inserts a TestStream primitive into a pipeline, taking a scope and a Config object and
+// producing an array of output PCollections.
+func TestStream(s beam.Scope, c Config) []beam.PCollection {
+	pyld := protox.MustEncode(c.CreatePayload())
+	outputs := []beam.FullType{c.ElmCoder.T}
+
+	outputMap := beam.External(s, urn, pyld, []beam.PCollection{}, outputs, false)
+
+	var ret []beam.PCollection
+	for _, val := range outputMap {
+		ret = append(ret, val)
+	}
+	return ret
```

Comment:
This should only ever be a single PCollection, so I was thinking this should be changed so that the return type is just a beam.PCollection, but didn't know if there was a clean looking way to do that.

Reply:
Agreed. Since implementations seem to all be single output collections, lets punt on tagged multi-collection map support for now. A later iteration can add a `CreateN` or `CreateMulti`call which would return a map.